File: python/enolib/reporters/reporter.py
from ..constants import (
    DOCUMENT,
    FIELD,
    FIELDSET,
    FIELDSET_ENTRY,
    LIST,
    LIST_ITEM,
    MULTILINE_FIELD_BEGIN,
    SECTION
)

DISPLAY = 'Display Line'
EMPHASIZE = 'Emphasize Line'
INDICATE = 'Indicate Line'
OMISSION = 'Insert Omission'
QUESTION = 'Question Line'

class Reporter:
    def __init__(self, context):
        self._context = context
        self._index = [None] * self._context.line_count
        self._snippet = [None] * self._context.line_count

        self._build_index()

    def _build_index(self):
        # TODO: Put this into the root class scope?
        def index_comments(element):
            if 'comments' in element:
                for comment in element['comments']:
                    self._index[comment['line']] = comment

        # TODO: Put this into the root class scope?
        def traverse(section):
            for element in section['elements']:
                index_comments(element)

                self._index[element['line']] = element

                if element['type'] == SECTION:
                    traverse(element)
                elif element['type'] == FIELD:
                    if 'continuations' in element:
                        for continuation in element['continuations']:
                            self._index[continuation['line']] = continuation
                elif element['type'] == MULTILINE_FIELD_BEGIN:
                    # Missing when reporting an unterminated multiline field
                    if 'end' in element:
                        self._index[element['end']['line']] = element['end']

                    if 'lines' in element:
                        for line in element['lines']:
                            self._index[line['line']] = line

                elif element['type'] == LIST:
                    if 'items' in element:
                        for item in element['items']:
                            index_comments(item)

                            self._index[item['line']] = item

                            if 'continuations' in item:
                                for continuation in item['continuations']:
                                    self._index[continuation['line']] = continuation

                elif element['type'] == FIELDSET:
                    if 'entries' in element:
                        for entry in element['entries']:
                            index_comments(entry)

                            self._index[entry['line']] = entry

                            if 'continuations' in entry:
                                for continuation in entry['continuations']:
                                    self._index[continuation['line']] = continuation

        traverse(self._context.document)

        for meta in self._context.meta:
            self._index[meta['line']] = meta

    def _tag_continuations(self, element, tag):
        scan_line = element['line'] + 1

        if not 'continuations' in element:
            return scan_line

        for continuation in element['continuations']:
            while scan_line < continuation['line']:
                self._snippet[scan_line] = tag
                scan_line += 1

            self._snippet[continuation['line']] = tag
            scan_line += 1

        return scan_line

    def _tag_continuables(self, element, collection, tag):
        scan_line = element['line'] + 1

        if not collection in element:
            return scan_line

        for continuable in element[collection]:
            while scan_line < continuable['line']:
                self._snippet[scan_line] = tag
                scan_line += 1

            self._snippet[continuable['line']] = tag

            scan_line = self._tag_continuations(continuable, tag)

        return scan_line

    def _tag_children(self, element, tag):
        if element['type'] == FIELD or element['type'] == LIST_ITEM or element['type'] == FIELDSET_ENTRY:
            return self._tag_continuations(element, tag)
        elif element['type'] == LIST:
            return self._tag_continuables(element, 'items', tag)
        elif element['type'] == FIELDSET and 'entries' in element:
            return self._tag_continuables(element, 'entries', tag)
        elif element['type'] == MULTILINE_FIELD_BEGIN:
            if 'lines' in element:
                for line in element['lines']:
                    self._snippet[line['line']] = tag

            if 'end' in element:
                self._snippet[element['end']['line']] = tag
                return element['end']['line'] + 1
            elif 'lines' in element:
                return element['lines'][-1]['line'] + 1
            else:
                return element['line'] + 1
        elif element['type'] == SECTION:
            return self._tag_section(element, tag)

    def _tag_section(self, section, tag, recursive=True):
        scan_line = section['line'] + 1

        for element in section['elements']:
            while scan_line < element['line']:
                self._snippet[scan_line] = tag
                scan_line += 1

            if not recursive and element['type'] == SECTION:
                break

            self._snippet[element['line']] = tag

            scan_line = self._tag_children(element, tag)

        return scan_line

    def indicate_line(self, element):
        self._snippet[element['line']] = INDICATE
        return self

    def question_line(self, element):
        self._snippet[element['line']] = QUESTION
        return self

    def report_comments(self, element):
        self._snippet[element['line']] = INDICATE
        for comment in element['comments']:
            self._snippet[comment['line']] = EMPHASIZE

        return self

    def report_element(self, element):
        self._snippet[element['line']] = EMPHASIZE
        self._tag_children(element, INDICATE)

        return self

    def report_elements(self, elements):
        for element in elements:
            self._snippet[element['line']] = EMPHASIZE
            self._tag_children(element, INDICATE)

        return self

    def report_line(self, instruction):
        self._snippet[instruction['line']] = EMPHASIZE

        return self

    def report_multiline_value(self, element):
        for line in element['lines']:
            self._snippet[line['line']] = EMPHASIZE

        return self

    def report_missing_element(self, parent):
        if parent['type'] != DOCUMENT:
            self._snippet[parent['line']] = INDICATE

        if parent['type'] == SECTION:
            self._tag_section(parent, QUESTION, False)
        else:
            self._tag_children(parent, QUESTION)

        return self

    def snippet(self):
        if all(tag is None for tag in self._snippet):
            for line in range(len(self._snippet)):
                self._snippet[line] = QUESTION
        else:
            # TODO: Possibly better algorithm for this

            for line, tag in enumerate(self._snippet):
                if tag is not None:
                    continue

                if(line + 2 < self._context.line_count and self._snippet[line + 2] is not None and self._snippet[line + 2] != DISPLAY or
                   line - 2 > 0 and self._snippet[line - 2] is not None and self._snippet[line - 2] != DISPLAY or
                   line + 1 < self._context.line_count and self._snippet[line + 1] is not None and self._snippet[line + 1] != DISPLAY or
                   line - 1 > 0 and self._snippet[line - 1] is not None and self._snippet[line - 1] != DISPLAY):
                    self._snippet[line] = DISPLAY
                elif line + 3 < self._context.line_count and self._snippet[line + 3] is not None and self._snippet[line + 3] != DISPLAY:
                    self._snippet[line] = OMISSION

            if self._snippet[-1] is None:
                self._snippet[-1] = OMISSION

        return self._print()

File: ruby/lib/enolib/reporters/reporter.rb
# frozen_string_literal: true

module Enolib
  class Reporter
    def initialize(context)
      @context = context
      @index = Array.new(@context.line_count)
      @snippet = Array.new(@context.line_count)

      build_index
    end

    def indicate_line(element)
      @snippet[element[:line]] = :indicate
      self
    end

    def question_line(element)
      @snippet[element[:line]] = :question
      self
    end

    def report_comments(element)
      @snippet[element[:line]] = :indicate
      element[:comments].each do |comment|
        @snippet[comment[:line]] = :emphasize
      end

      self
    end

    def report_element(element)
      @snippet[element[:line]] = :emphasize
      tag_element(element, :indicate)

      self
    end

    def report_elements(elements)
      elements.each do |element|
        @snippet[element[:line]] = :emphasize
        tag_element(element, :indicate)
      end

      self
    end

    def report_line(instruction)
      @snippet[instruction[:line]] = :emphasize

      self
    end

    def report_multiline_value(element)
      element[:lines].each do |line|
        @snippet[line[:line]] = :emphasize
      end

      self
    end

    def report_missing_element(parent)
      @snippet[parent[:line]] = :indicate unless parent[:type] == :document

      if parent[:type] == :section
        tag_section(parent, :question, false)
      else
        tag_element(parent, :question)
      end

      self
    end

    def snippet
      if @snippet.none?
        (0...@snippet.length).each do |line|
          @snippet[line] = :question
        end
      else
        # TODO: Possibly better algorithm for this
        @snippet.each_with_index do |tag, line|
          next if tag

          if line + 2 < @context.line_count && @snippet[line + 2] && @snippet[line + 2] != :display ||
             line - 2 > 0 && @snippet[line - 2] && @snippet[line - 2] != :display ||
             line + 1 < @context.line_count && @snippet[line + 1] && @snippet[line + 1] != :display ||
             line - 1 > 0 && @snippet[line - 1] && @snippet[line - 1] != :display
            @snippet[line] = :display
          elsif line + 3 < @context.line_count && @snippet[line + 3] && @snippet[line + 3] != :display
            @snippet[line] = :omission
          end
        end

        @snippet[-1] = :omission unless @snippet[-1]
      end

      print
    end

    private

    def index_comments(element)
      return unless element.has_key?(:comments)

      element[:comments].each do |comment|
        @index[comment[:line]] = comment
      end
    end

    def traverse(section)
      section[:elements].each do |element|
        index_comments(element)

        @index[element[:line]] = element

        if element[:type] == :section
          traverse(element)
        elsif element[:type] == :field
          if element.has_key?(:continuations)
            element[:continuations].each do |continuation|
              @index[continuation[:line]] = continuation
            end
          end
        elsif element[:type] == :multiline_field_begin
          # Missing when reporting an unterminated multiline field
          if element.has_key?(:end)
            @index[element[:end][:line]] = element[:end]
          end

          if element.has_key?(:lines)
            element[:lines].each do |line|
              @index[line[:line]] = line
            end
          end
        elsif element[:type] == :list
          if element.has_key?(:items)
            element[:items].each do |item|
              index_comments(item)

              @index[item[:line]] = item

              if item.has_key?(:continuations)
                item[:continuations].each do |continuation|
                  @index[continuation[:line]] = continuation
                end
              end
            end
          end
        elsif element[:type] == :fieldset
          if element.has_key?(:entries)
            element[:entries].each do |entry|
              index_comments(entry)

              @index[entry[:line]] = entry

              if entry.has_key?(:continuations)
                entry[:continuations].each do |continuation|
                  @index[continuation[:line]] = continuation
                end
              end
            end
          end
        end
      end
    end

    def build_index
      traverse(@context.document)

      @context.meta.each do |instruction|
        @index[instruction[:line]] = instruction
      end
    end

    def tag_continuations(element, tag)
      scan_line = element[:line] + 1

      return scan_line unless element.has_key?(:continuations)

      element[:continuations].each do |continuation|
        while scan_line < continuation[:line]
          @snippet[scan_line] = tag
          scan_line += 1
        end

        @snippet[continuation[:line]] = tag
        scan_line += 1
      end

      scan_line
    end

    def tag_continuables(element, collection, tag)
      scan_line = element[:line] + 1

      return scan_line unless element.has_key?(collection)

      element[collection].each do |continuable|
        while scan_line < continuable[:line]
          @snippet[scan_line] = tag
          scan_line += 1
        end

        @snippet[continuable[:line]] = tag

        scan_line = tag_continuations(continuable, tag)
      end

      scan_line
    end

    def tag_element(element, tag)
      if element[:type] == :field || element[:type] == :list_item || element[:type] == :fieldset_entry
        return tag_continuations(element, tag)
      elsif element[:type] == :list
        return tag_continuables(element, :items, tag)
      elsif element[:type] == :fieldset && element.has_key?(:entries)
        return tag_continuables(element, :entries, tag)
      elsif element[:type] == :multiline_field_begin
        if element.has_key?(:lines)
          element[:lines].each do |line|
            @snippet[line[:line]] = tag
          end
        end

        if element.has_key?(:end)
          @snippet[element[:end][:line]] = tag
          return element[:end][:line] + 1
        elsif element.has_key?(:lines)
          return element[:lines][-1][:line] + 1
        else
          return element[:line] + 1
        end
      elsif element[:type] == :section
        return tag_section(element, tag)
      end
    end

    def tag_section(section, tag, recursive=true)
      scan_line = section[:line] + 1

      section[:elements].each do |element|
        while scan_line < element[:line]
          @snippet[scan_line] = tag
          scan_line += 1
        end

        break if element[:type] == :section && !recursive

        @snippet[element[:line]] = tag

        scan_line = tag_element(element, tag)
      end

      scan_line
    end
  end
end
